tempesta_fw/http.c

```diff
@@ -383,18 +369,636 @@ tfw_http_send_504(TfwHttpReq *req, const char *reason)
+ * is not re-sent or re-scheduled by default. Configuration option
+ * can be used to have that request re-sent or re-scheduled.
+ *
+ * As forwarding is paused after a non-idempotent request is sent,
+ * there can be only one such request among forwarded requests, and
+ * that's @srv_conn->msg_sent.
+ *
+ * Note: @srv_conn->msg_sent may change in result.
+ */
+static inline void
+tfw_http_conn_treatnip(TfwSrvConn *srv_conn, struct list_head *equeue)
+{
+	TfwServer *srv = (TfwServer *)srv_conn->peer;
+	TfwHttpReq *req_sent = (TfwHttpReq *)srv_conn->msg_sent;
+
+	if (req_sent && tfw_http_req_is_nip(req_sent)
+	    && likely(!(srv->sg->flags & TFW_SRV_RETRY_NIP)))
+	{
+		BUG_ON(list_empty(&req_sent->nip_list));
+		srv_conn->msg_sent =
+			/* list_is_first(&req_sent->fwd_list, fwd_queue); */
+			(srv_conn->fwd_queue.next == &req_sent->fwd_list) ?
+			NULL : (TfwMsg *)list_prev_entry(req_sent, fwd_list);
+		tfw_http_req_move2equeue(srv_conn, req_sent, equeue, 504);
+	}
+}
+
+/*
+ * Re-forward requests in a server connection. Requests that exceed
+ * the set limits are evicted.
+ */
+static TfwHttpReq *
+tfw_http_conn_resend(TfwSrvConn *srv_conn, bool first, struct list_head *equeue)
+{
+	TfwHttpReq *req, *tmp, *req_resent = NULL;
+	TfwServer *srv = (TfwServer *)srv_conn->peer;
+	struct list_head *end, *fwd_queue = &srv_conn->fwd_queue;
+
+	TFW_DBG2("%s: conn=[%p] one_msg=[%s]\n",
+		 __func__, srv_conn, first ? "true" : "false");
+	BUG_ON(!srv_conn->msg_sent);
+	BUG_ON(list_empty(&((TfwHttpReq *)srv_conn->msg_sent)->fwd_list));
+
+	req = list_first_entry(fwd_queue, TfwHttpReq, fwd_list);
+	end = ((TfwHttpReq *)srv_conn->msg_sent)->fwd_list.next;
+
+	/* Similar to list_for_each_entry_safe_from() */
+	for (tmp = list_next_entry(req, fwd_list);
+	     &req->fwd_list != end;
+	     req = tmp, tmp = list_next_entry(tmp, fwd_list))
+	{
+		if (tfw_http_req_evict_timeout(srv_conn, srv, req, equeue))
+			continue;
+		if (tfw_http_req_evict_retries(srv_conn, srv, req, equeue))
+			continue;
+		if (!tfw_http_req_fwd_send(srv_conn, srv, req, equeue))
+			continue;
+		req_resent = req;
+		if (unlikely(first))
+			break;
+	}
+
+	return req_resent;
+}
+
+/*
+ * Re-send only the first unanswered request in the forwarding queue.
+ */
+static inline TfwHttpReq *
+tfw_http_conn_resend_first(TfwSrvConn *srv_conn, struct list_head *equeue)
+{
+	return tfw_http_conn_resend(srv_conn, true, equeue);
+}
+
+/*
+ * Re-send all unanswered requests in the forwarding queue.
+ */
+static inline TfwHttpReq *
+tfw_http_conn_resend_all(TfwSrvConn *srv_conn, struct list_head *equeue)
+{
+	return tfw_http_conn_resend(srv_conn, false, equeue);
+}
+
+/*
+ * Handle the complete re-forwarding of requests in a server connection
+ * that is being repaired, after the first request had been re-forwarded.
+ * The connection is not scheduled until all requests in it are re-sent.
+ */
+static void
+__tfw_http_conn_fwd_repair(TfwSrvConn *srv_conn, struct list_head *equeue)
```

Comment:
I tried to do that, but it lead to really, really long function names. If the HTTP client and server logic is separated in the future, then it will be clear which function is what simply by the location of a function. Only external functions will need to be named explicitly.

I would leave the names as they are for now. What do you think?